add docs

claws opened the pull request:
This pull request adds documentation to the uvloop project. A live demonstration can be found [here](http://uvloop-demo.readthedocs.io/en/docs/). The documentation uses Sphinx and can be linked to [readthedocs](https://readthedocs.org/) where many Python packages host their documentation.

The documentation contains a main page containing a brief overview. Then there is a user guide (install and using uvloop), a developers guide (getting the source, building, testing, docs) and the API page. The makefile has been updated to make it easy to build the docs.

I have also updated the loop.pyx file with comments similar to their asyncio counterparts. This file also contains a compiler directive (embedsignature) that allows Sphinx to pull the docs trings from Cython modules.

This work builds upon an earlier PR (which I closed because I stuffed up the git repo with a rebase that went haywire) in which there were some existing review comments. The content of this PR has addressed some of these comments, a summary is presented below:

- Added requirements.dev.txt to contain Cython.
  - if you think the project should use pytest for testing should the file also contain pytest?
- Added a link to the ReadTheDocs docs in the README.rst.
- Added a link to the magic.io uvloop blog post in the docs index page.
- Added a reference for using pytest as well as unittest.
- Added a sentence to the dev docs stating that any deviation from the reference `asyncio` behavior is considered a bug.
- Removed reference to needing Cython for user installs.
- Added ``install Cython`` note to dev docs.
- Sphinx docs/projects typically use a logo. This project does not have a logo yet.
  - one consequence is that the docs theme will use the title `uvloop` if no image is provided. It looks a bit clumsy but if you add a logo image then the title can be muted so it should be fine.
- embedsignature is used to allow sphinx to pull docs strings from the Cython modules.

1st1 commented on docs/index.rst:
```diff
@@ -0,0 +1,66 @@
+
+.. raw:: html
```
Let's just copy the shields code from README.rst. No need to center them.

claws replied:
Back to two shields.

1st1 commented on docs/user/index.rst:
```diff
@@ -0,0 +1,40 @@
+==========
+
+This section of the documentation provides information about how to install
+uvloop as well as how to use uvloop.
```
I'd rephrase to:

    This section of the documentation provides information about 
    how to use uvloop.

1st1 commented on docs/conf.py:
```diff
@@ -0,0 +1,303 @@
+#!/usr/bin/env python3
```
Can we remove all the commented code etc?

claws replied:
Done

1st1 commented:
Another thought. I'd like this to be split in two PRs: one that adds docstrings to the Loop methods; and one that sets up docs/sphinx/etc.

1st1 commented on docs/dev/index.rst:
```diff
@@ -0,0 +1,175 @@
+        $ python3 -m venv myvenv
+        $ cd myvenv
+        $ source bin/activate
+        $ cd ..
```
Let's add a note about `pip -e .`

claws replied:
Added it down in the tests section.

1st1 commented on docs/dev/index.rst:
```diff
@@ -0,0 +1,175 @@
+Alternatively, you can run the tests from the ``tests`` directory using the
+standard library `unittest` approach. When using this approach you must
+specify the ``PYTHONPATH`` so that the `uvloop` package can be resolved from
+down in the ``tests`` directory.
```
Again, `pip -e` solves the `PYTHONPATH` problem.

1st1 commented on docs/dev/index.rst:
```diff
@@ -0,0 +1,175 @@
+Then open http://localhost:8000 in a browser and navigate to
+``_build/html/index.html``.
+
+.. note::
```
I don't think we need to go into too much detail here... Let's keep the docs as concise as possible ;)

1st1 commented on requirements.dev.txt:
```diff
@@ -0,0 +1,2 @@
+Cython
```
Let's specify the min required versions of Cython and sphinx here.

1st1 commented:
Alright, I've left some feedback. Generally looks good, aside from a few nits. Let's break this PR into two though.

1st1 commented on docs/Makefile:
```diff
@@ -0,0 +1,192 @@
+# Makefile for Sphinx documentation
```
Maybe we don't need this Makefile?  Would it be possible to just call sphinx directly (using `python -m sphinx`) from the `uvloop/Makefile`?

claws commented:
I think that this addresses all the review comments.

- removed the loop.pyx comments update to a separate PR.
- loop.pyx now contains just the Sphinx related change supporting docs.
- same shields as README
- updated user doc with suggested re-write.
- trimmed Sphinx conf.py to minimal content (removed commented out stuff)
- added reference to using ``pip install -e .`` instead of using PYTHONPATH.
- trimmed testing section
- trimmed the docs section as suggested.
- updated requirements.txt with package versions
- Removed Sphinx Makefile with a call in from the main Makefile directly to
  the sphinx-build call.

1st1 commented:
Looks good! Will merge it today.

1st1 commented:
Committed in 8511fec1f9a610b0b26547a1620ce034ad9d9947. Thanks a lot!
